In [2]:
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

import numpy as np

In [82]:
df = pd.read_csv('data/titanic.csv')
X = df.drop(['Survived', 'Name'], axis=1)
y = df.Survived

Pclass     Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
0         3    male  22.0                        1                        0   
1         1  female  38.0                        1                        0   
2         3  female  26.0                        0                        0   
3         1  female  35.0                        1                        0   
4         3    male  35.0                        0                        0   
..      ...     ...   ...                      ...                      ...   
882       2    male  27.0                        0                        0   
883       1  female  19.0                        0                        0   
884       3  female   7.0                        1                        2   
885       1    male  26.0                        0                        0   
886       3    male  32.0                        0                        0   

        Fare  
0     7.2500  
1    71.2833  
2     7.9250  
3    53.1000  
4     8.0500  
..       ...  
882  13.0000  
883  30.0000  
884  23.4500  
885  30.0000  
886   7.7500  

[887 rows x 6 columns]

In [46]:
onehot = OneHotEncoder(sparse_output=False, drop='first')

X_is_objects = onehot.fit_transform(X.select_dtypes(include=['object']))
X_not_is_objects = X.select_dtypes(exclude=['object'])

X_all = np.append(X_not_is_objects, X_is_objects, axis=1)

In [47]:
X_train, X_test, y_train, y_test = train_test_split(X_all, y, train_size=0.8, random_state=42)

knn = KNeighborsClassifier(n_neighbors=8)
knn.fit(X_train, y_train)

accuracy_score(y_test, knn.predict(X_test))

0.702247191011236

Resultado de 68.53% sem normalização dos dados

Subindo para 70.22% ao tratar a coluna 'Sex' como binário

Subindo para 75.84% removendo a coluna 'Fare'

In [48]:
normalizer = MinMaxScaler()
X_norm = pd.DataFrame(
    columns=['Pclass', 'Age', 'Siblings/Spouses Aboard', 'Parents/Children Aboard', 'Fare', 'Sex'],
    data=normalizer.fit_transform(X_all))
X_norm

Pclass       Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
0       1.0  0.271174                    0.125                 0.000000   
1       0.0  0.472229                    0.125                 0.000000   
2       1.0  0.321438                    0.000                 0.000000   
3       0.0  0.434531                    0.125                 0.000000   
4       1.0  0.434531                    0.000                 0.000000   
..      ...       ...                      ...                      ...   
882     0.5  0.334004                    0.000                 0.000000   
883     0.0  0.233476                    0.000                 0.000000   
884     1.0  0.082684                    0.125                 0.333333   
885     0.0  0.321438                    0.000                 0.000000   
886     1.0  0.396833                    0.000                 0.000000   

         Fare  Sex  
0    0.014151  1.0  
1    0.139136  0.0  
2    0.015469  0.0  
3    0.103644  0.0  
4    0.015713  1.0  
..        ...  ...  
882  0.025374  1.0  
883  0.058556  0.0  
884  0.045771  0.0  
885  0.058556  1.0  
886  0.015127  1.0  

[887 rows x 6 columns]

In [49]:
X_train, X_test, y_train, y_test = train_test_split(X_norm, y, train_size=0.8, random_state=42)

knn.fit(X_train, y_train)
accuracy_score(y_test, knn.predict(X_test))

0.7696629213483146

Resultado de 70.78% COM normalização dos dados

Subindo para 76.96% ao tratar a coluna 'Sex' como binário

Subindo para 77.52% removendo a coluna 'Fare'

Porém, no fim, tratando com um bom número de vizinhos (4) a acuracidade se mantém ligeiramente maior ao tratar com a coluna 'Fare'

In [51]:
max_value_accuracy_score = []

for x in range(8):
    knn = KNeighborsClassifier(n_neighbors=x+1)
    knn.fit(X_train, y_train)
    score = accuracy_score(y_test, knn.predict(X_test))
    max_value_accuracy_score.append(score)
    print(x+1, score)

print(f'{max(max_value_accuracy_score) * 100:.2f} % com 4 vizinhos')

1 0.7528089887640449
2 0.7640449438202247
3 0.7921348314606742
4 0.8089887640449438
5 0.797752808988764
6 0.7921348314606742
7 0.7696629213483146
8 0.7696629213483146
80.90 % com 4 vizinhos


In [83]:
X

Pclass     Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
0         3    male  22.0                        1                        0   
1         1  female  38.0                        1                        0   
2         3  female  26.0                        0                        0   
3         1  female  35.0                        1                        0   
4         3    male  35.0                        0                        0   
..      ...     ...   ...                      ...                      ...   
882       2    male  27.0                        0                        0   
883       1  female  19.0                        0                        0   
884       3  female   7.0                        1                        2   
885       1    male  26.0                        0                        0   
886       3    male  32.0                        0                        0   

        Fare  
0     7.2500  
1    71.2833  
2     7.9250  
3    53.1000  
4     8.0500  
..       ...  
882  13.0000  
883  30.0000  
884  23.4500  
885  30.0000  
886   7.7500  

[887 rows x 6 columns]

In [86]:
#Acertou
Carrie_Constance_Chaffee_1 = pd.DataFrame({
'Pclass': [1],
'Sex':[0],
'Age':[47],
'Siblings/Spouses Aboard':[1],
'Parents/Children Aboard':[0],
'Fare':[61],
})

#Errou
Herbert_Fuller_Chaffee_0 = pd.DataFrame({
'Pclass': [1],
'Sex':[1],
'Age':[46],
'Siblings/Spouses Aboard':[1],
'Parents/Children Aboard':[0],
'Fare':[61],
})

df_final = pd.DataFrame(
    columns=['Pclass', 'Age', 'Siblings/Spouses Aboard', 'Parents/Children Aboard', 'Fare', 'Sex']
)

X = pd.concat([Carrie_Constance_Chaffee_1, Herbert_Fuller_Chaffee_0, X], ignore_index=True)
X = X.drop('')

# df_final = normalizer.fit_transform(df_final)
# print(df_final)

# X_train, X_test, y_train, y_test = train_test_split(df_final, y, train_size=0.8, random_state=42)

# knn.fit(X_train, y_train)
# accuracy_score(y_test, knn.predict(X_test))

# print(accuracy_score([1], knn.predict(Carrie_Constance_Chaffee_1)))
# print(accuracy_score([1], knn.predict(Herbert_Fuller_Chaffee_0)))

Pclass     Sex   Age  Siblings/Spouses Aboard  Parents/Children Aboard  \
0         1       0  47.0                        1                        0   
1         1       1  46.0                        1                        0   
2         1       0  47.0                        1                        0   
3         1       1  46.0                        1                        0   
4         1       0  47.0                        1                        0   
..      ...     ...   ...                      ...                      ...   
888       2    male  27.0                        0                        0   
889       1  female  19.0                        0                        0   
890       3  female   7.0                        1                        2   
891       1    male  26.0                        0                        0   
892       3    male  32.0                        0                        0   

      Fare  Survived                            Name  
0    61.00       NaN                             NaN  
1    61.00       NaN                             NaN  
2    61.00       NaN                             NaN  
3    61.00       NaN                             NaN  
4    61.00       NaN                             NaN  
..     ...       ...                             ...  
888  13.00       0.0            Rev. Juozas Montvila  
889  30.00       1.0     Miss. Margaret Edith Graham  
890  23.45       0.0  Miss. Catherine Helen Johnston  
891  30.00       1.0            Mr. Karl Howell Behr  
892   7.75       0.0              Mr. Patrick Dooley  

[893 rows x 8 columns]

In [ ]:
Jose_Joaquim_de_Brito_0 = pd.DataFrame({
'Pclass': [2],
'Age':[41],
'Siblings/Spouses Aboard':[0],
'Parents/Children Aboard':[0],
'Fare':[13],
'Sex':[1],
})

Manuel_Goncalves_Estanislau_0 = pd.DataFrame({
'Pclass': [3],
'Age':[37],
'Siblings/Spouses Aboard':[0],
'Parents/Children Aboard':[0],
'Fare':[13],
'Sex':[1],
})

Rev_Jhon_Stuart_Holden_NotGo = pd.DataFrame({
'Pclass': [1],
'Age':[40],
'Siblings/Spouses Aboard':[1],
'Parents/Children Aboard':[0],
'Fare':[13.050881],
'Sex':[1],
})

Jose_Pedro_CarrauEsteves_0 = pd.DataFrame({
'Pclass': [1],
'Age':[17],
'Siblings/Spouses Aboard':[1],
'Parents/Children Aboard':[0],
'Fare':[47],
'Sex':[1],
})
